src/packet/sym_encrypted_protected_data.rs

```diff
@@ -53,26 +52,105 @@ impl SymEncryptedProtectedData {
     }
 
     /// Encrypts the data using the given symmetric key.
-    pub fn encrypt_with_rng<R: CryptoRng + Rng>(
+    pub fn encrypt_seipdv1<R: CryptoRng + Rng>(
         rng: R,
         alg: SymmetricKeyAlgorithm,
         key: &[u8],
         plaintext: &[u8],
     ) -> Result<Self> {
-        let data = alg.encrypt_protected_with_rng(rng, key, plaintext)?;
+        let data = alg.encrypt_protected(rng, key, plaintext)?;
 
         Ok(SymEncryptedProtectedData {
             packet_version: Default::default(),
             data: Data::V1 { data },
         })
     }
 
-    /// Same as [`encrypt_with_rng`], but uses [`thread_rng`] for RNG.
-    ///
-    /// [`encrypt_with_rng`]: SymEncryptedProtectedData::encrypt_with_rng
-    /// [`thread_rng`]: rand::thread_rng
-    pub fn encrypt(alg: SymmetricKeyAlgorithm, key: &[u8], plaintext: &[u8]) -> Result<Self> {
-        Self::encrypt_with_rng(&mut thread_rng(), alg, key, plaintext)
+    /// Encrypts the data using the given symmetric key.
+    pub fn encrypt_seipdv2<R: CryptoRng + Rng>(
+        mut rng: R,
+        sym_alg: SymmetricKeyAlgorithm,
+        aead: AeadAlgorithm,
+        chunk_size: u8,
+        session_key: &[u8],
+        plaintext: &[u8],
+    ) -> Result<Self> {
+        // TODO: DRY decrypt/encrypt code duplication?
+
+        // Initial key material is the session key.
+        let ikm = session_key;
+
+        // Generate new salt for this seipd packet.
+        let mut salt = [0u8; 32];
+        rng.fill(&mut salt[..]);
+
+        let info = [
+            Tag::SymEncryptedProtectedData.encode(), // packet type
+            0x02,                                    // version
+            sym_alg.into(),
+            aead.into(),
+            chunk_size,
+        ];
+
+        let chunk_size_expanded = expand_chunk_size(chunk_size);
+        let hk = hkdf::Hkdf::<Sha256>::new(Some(&salt[..]), ikm);
+        let mut okm = [0u8; 42];
+        hk.expand(&info, &mut okm).expect("42");
+
+        let message_key = &okm[..sym_alg.key_size()];
+        let raw_iv_len = aead.nonce_size() - 8;
+        let iv = &okm[sym_alg.key_size()..sym_alg.key_size() + raw_iv_len];
+        let mut nonce = vec![0u8; aead.nonce_size()];
+        nonce[..raw_iv_len].copy_from_slice(iv);
+
+        let mut out = Vec::new();
```

Comment:
can we preallocate this with a known size?

Reply:
Yes. I think I've done the right math in e18252c891bc988701a30508577cfd3b5ce1fcd5.

It's just a bit annoying that this calculation is pretty convoluted and takes up a lot of space (putting cognitive load on future readers), so I tried my best to comment what's going on.

(But if we switch to streaming processing, that whole allocation will go away, anyway :smiley:)

Reply:
(I'll do the equivalent preallocation for the decryption allocation, in the DRY [comment above](https://github.com/rpgp/rpgp/pull/392#discussion_r1760973235))

Reply:
I added a `debug_assert_eq` that the allocation length calculation matches the final len of `out`.

I'm unsure if that's reasonable. Nut it seemed silly if the preallocation calculation was wrong, and we had no chance to ever notice. Would you find a WARN (?) log output appropriate for this? (Rather than panicing in debug mode)